File: MapsterEvents.Service/Services/CategoryService.cs
using Mapster;
using MapsterEvents.Core.DTOs;
using MapsterEvents.Core.Entities;
using MapsterEvents.Core.Interfaces;
using MapsterEvents.Repository.Data;

namespace MapsterEvents.Service.Services
{
    /// <summary>
    /// Kategori servis implementasyonu
    /// </summary>
    public class CategoryService : ICategoryService
    {
        private readonly ICategoryRepository _categoryRepository;
        private readonly MapsterEventsDbContext _context;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="categoryRepository">Kategori repository</param>
        /// <param name="context">Veritabanı context (Unit of Work için)</param>
        public CategoryService(ICategoryRepository categoryRepository, MapsterEventsDbContext context)
        {
            _categoryRepository = categoryRepository;
            _context = context;
        }

        /// <summary>
        /// Tüm kategorileri getirir
        /// </summary>
        /// <returns>Kategori listesi</returns>
        public async Task<IEnumerable<CategoryDto>> GetAllCategoriesAsync()
        {
            var categories = await _categoryRepository.GetAllAsync();
            return categories.Adapt<IEnumerable<CategoryDto>>();
        }

        /// <summary>
        /// Kategori detayını getirir
        /// </summary>
        /// <param name="categoryId">Kategori ID'si</param>
        /// <returns>Kategori detayı</returns>
        public async Task<CategoryDto?> GetCategoryByIdAsync(int categoryId)
        {
            var category = await _categoryRepository.GetByIdAsync(categoryId);
            return category?.Adapt<CategoryDto>();
        }

        /// <summary>
        /// Kategorileri etkinlik sayısıyla birlikte getirir
        /// </summary>
        /// <returns>Kategori listesi (EventCount hesaplanmış)</returns>
        public async Task<IEnumerable<CategoryDto>> GetCategoriesWithEventCountAsync()
        {
            // Repository'den eager loaded entities'leri al
            var categories = await _categoryRepository.GetAllWithEventsAsync();
            
            // Service layer'da Entity -> DTO mapping (EventCount hesaplama dahil)
            return categories.Adapt<IEnumerable<CategoryDto>>();
        }

        /// <summary>
        /// Yeni kategori oluşturur
        /// </summary>
        /// <param name="categoryCreateDto">Kategori oluşturma DTO'su</param>
        /// <returns>Oluşturulan kategori</returns>
        public async Task<CategoryDto> CreateCategoryAsync(CategoryCreateDto categoryCreateDto)
        {
            // İş kuralı: Kategori adı benzersiz olmalı
            var isNameInUse = await _categoryRepository.IsCategoryNameInUseAsync(categoryCreateDto.Name);
            if (isNameInUse)
            {
                throw new InvalidOperationException("Bu kategori adı zaten kullanılıyor");
            }

            var category = categoryCreateDto.Adapt<Category>();
            var createdCategory = await _categoryRepository.AddAsync(category);
            
            // Unit of Work pattern - Tek transaction içinde save
            await _context.SaveChangesAsync();
            
            return createdCategory.Adapt<CategoryDto>();
        }

        /// <summary>
        /// Kategori günceller
        /// </summary>
        /// <param name="categoryId">Kategori ID'si</param>
        /// <param name="categoryUpdateDto">Kategori güncelleme DTO'su</param>
        /// <returns>Güncellenen kategori</returns>
        public async Task<CategoryDto?> UpdateCategoryAsync(int categoryId, CategoryUpdateDto categoryUpdateDto)
        {
            var existingCategory = await _categoryRepository.GetByIdAsync(categoryId);
            if (existingCategory == null)
            {
                return null;
            }

            // İş kuralı: Kategori adı benzersiz olmalı (mevcut kategori hariç)
            var isNameInUse = await _categoryRepository.IsCategoryNameInUseAsync(categoryUpdateDto.Name, categoryId);
            if (isNameInUse)
            {
                throw new InvalidOperationException("Bu kategori adı zaten kullanılıyor");
            }

            // Mapster ile DTO'dan entity'ye değerleri kopyala
            categoryUpdateDto.Adapt(existingCategory);

            var updatedCategory = _categoryRepository.UpdateAsync(existingCategory);
            
            // Unit of Work pattern - Tek transaction içinde save
            await _context.SaveChangesAsync();
            
            return updatedCategory.Adapt<CategoryDto>();
        }

        /// <summary>
        /// Kategori siler
        /// </summary>
        /// <param name="categoryId">Kategori ID'si</param>
        /// <returns>Başarılı mı</returns>
        public async Task<bool> DeleteCategoryAsync(int categoryId)
        {
            var existingCategory = await _categoryRepository.GetByIdAsync(categoryId);
            if (existingCategory == null)
            {
                return false;
            }

            // İş kuralı: İçinde etkinlik olan kategori silinemez
            var hasEvents = await _categoryRepository.HasEventsAsync(categoryId);
            if (hasEvents)
            {
                throw new InvalidOperationException("İçinde etkinlik olan kategori silinemez");
            }

            await _categoryRepository.DeleteAsync(existingCategory);
            
            // Unit of Work pattern - Tek transaction içinde save
            await _context.SaveChangesAsync();
            
            return true;
        }

        /// <summary>
        /// Kategori adı kullanımda mı kontrol eder
        /// </summary>
        /// <param name="name">Kategori adı</param>
        /// <param name="excludeId">Hariç tutulacak ID</param>
        /// <returns>Kullanımda mı</returns>
        public async Task<bool> IsCategoryNameInUseAsync(string name, int? excludeId = null)
        {
            return await _categoryRepository.IsCategoryNameInUseAsync(name, excludeId);
        }
    }
}

File: MapsterEvents.UI/src/app/services/category.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Category } from '../models/category.model';
import { environment } from '../../environments/environment';

/**
 * Kategori servisi
 * Created by Hamza Canturk - https://hamzacanturk.com/
 */
@Injectable({
  providedIn: 'root'
})
export class CategoryService {
  private apiUrl = `${environment.apiUrl}/categories`;

  constructor(private http: HttpClient) { }

  /**
   * Tüm kategorileri getirir
   */
  getCategories(): Observable<Category[]> {
    return this.http.get<Category[]>(this.apiUrl);
  }

  /**
   * Kategori detayını getirir
   */
  getCategoryById(id: number): Observable<Category> {
    return this.http.get<Category>(`${this.apiUrl}/${id}`);
  }

  /**
   * Kategorileri etkinlik sayısıyla birlikte getirir
   */
  getCategoriesWithEventCount(): Observable<Category[]> {
    return this.http.get<Category[]>(`${this.apiUrl}/with-event-count`);
  }

  /**
   * Yeni kategori oluşturur
   */
  createCategory(category: { name: string; description: string }): Observable<Category> {
    return this.http.post<Category>(this.apiUrl, category);
  }

  /**
   * Kategori günceller
   */
  updateCategory(id: number, category: { id: number; name: string; description: string }): Observable<Category> {
    return this.http.put<Category>(`${this.apiUrl}/${id}`, category);
  }

  /**
   * Kategori siler
   */
  deleteCategory(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  /**
   * Kategori adının kullanımda olup olmadığını kontrol eder
   */
  checkCategoryName(name: string, excludeId?: number): Observable<boolean> {
    let url = `${this.apiUrl}/check-name?name=${encodeURIComponent(name)}`;
    if (excludeId) {
      url += `&excludeId=${excludeId}`;
    }
    return this.http.get<boolean>(url);
  }
}